### ISSUE TITLE
[java] False positives for Cloneable rules

### ISSUE THREAD

**Issue by Fernal73** [pmd/pmd#3104]
**Affects PMD Version:** 
6.31.0

**Rule(s):**
[CloneMethodMustImplementCloneable](https://pmd.github.io/latest/pmd_rules_java_errorprone.html#clonemethodmustimplementcloneable)
[CloneThrowsCloneNotSupportedException](https://pmd.github.io/latest/pmd_rules_java_errorprone.html#clonethrowsclonenotsupportedexception)
[CloneReturnTypeMustMatchClassName](https://pmd.github.io/latest/pmd_rules_java_errorprone.html#clonemethodreturntypemustmatchclassname)

**Description:**
False positives for above three rules are reported as follows( ignore line numbers):

```text
DSAlgos/src/main/java/ds/ITreeNode.java:36:	CloneMethodMustImplementCloneable:	clone() method should be implemented only if implementing Cloneable interface

DSAlgos/src/main/java/ds/ITreeNode.java:36:	CloneThrowsCloneNotSupportedException:	clone() method should throw CloneNotSupportedException

DSAlgos/src/main/java/ds/TreeNode.java:209:	CloneMethodReturnTypeMustMatchClassName:	The return type of the clone() method must be the class name when implements Cloneable

DSAlgos/src/main/java/ds/TreeNode.java:209:	CloneThrowsCloneNotSupportedException:	clone() method should throw CloneNotSupportedException

```

**Code Sample demonstrating the issue:**

```java
package ds; 
interface ITreeNode<E> extends Cloneable { 
ITreeNode<E> clone(); 
}

package ds; 
public class TreeNode<T extends Comparable<T>> implements ITreeNode<T> { 
@Override public ITreeNode<T> clone() { 
try { 
return (ITreeNode<T>) super.clone();
 } catch (CloneNotSupportedException cnse) {
 throw new AssertionError("Shouldn't reach here...", cnse); }
 } 
}
```

**Expected outcome:**

*   No violations must be reported.

**Running PMD through:** *[CLI | Ant | Maven | Gradle | Designer | Other]*

First posted [here](https://github.com/pmd/pmd/discussions/3103).

**Comment by oowekyala** [comment]
The only false positive I see here is that of `CloneMethodMustImplementCloneable`, though I think this wouldn't be a problem with a complete classpath

The others are valid violations. Whether those rules make sense is another matter. I'm not sure `CloneThrowsCloneNotSupportedException` is useful, as calling `super.clone()` already forces you to declare the exception or handle it... why does it exist?

**Comment by Fernal73** [comment]
@oowekyala I was quite sure I used to False-positive template. Kindly check.

**Comment by Fernal73** [comment]
> 
> The others are valid violations. Whether those rules make sense is another matter. I'm not sure `CloneThrowsCloneNotSupportedException` is useful, as calling `super.clone()` already forces you to declare the exception or handle it... why does it exist?
> 

The rules don't handle interfaces well.

**Comment by Fernal73** [comment]
> The only false positive I see here is that of `CloneMethodMustImplementCloneable`, though I think this wouldn't be a problem with a complete classpath
> 

I assume this has something to do with auxclasspath needing to find the compiled classes.
I'm using Maven and this is my configuration file.

https://github.com/Fernal73/DSAlgos/blob/master/pom.xml

Give me a shout out if there's something to be updated in there. Maven can be huge headache, sometimes.

**Comment by oowekyala** [comment]
> @oowekyala I was quite sure I used the False-positive template. Kindly check.

Ah nevermind, I didn't think that template added the bug label. I always have to remove the bug label and add false-positive...

**Comment by Fernal73** [comment]
> > The only false positive I see here is that of `CloneMethodMustImplementCloneable`, though I think this wouldn't be a problem with a complete classpath
> 
> I assume this has something to do with auxclasspath needing to find the compiled classes.
> I'm using Maven and this is my configuration file.
> 
> https://github.com/Fernal73/DSAlgos/blob/master/pom.xml
> 
> Give me a shout out if there's something to be updated in there. Maven can be a huge headache, sometimes.

https://maven.apache.org/plugins/maven-pmd-plugin/pmd-mojo.html#typeResolution

Using the goal `pmd` with `typeResolution` set to true fixes this after the classes have been compiled.

**Comment by Fernal73** [comment]
> > The only false positive I see here is that of `CloneMethodMustImplementCloneable`, though I think this wouldn't be a problem with a complete classpath
> 
Updated the pom to explicitly specify typeResolution to be true (which it defaults to, anyway). I probably ran the PMD goal as standalone on the un-compiled sources. Hence, this violation.

**Comment by adangel** [comment]
> There should be an additional line:
> "Type resolution will not function properly possibly flagging erroneous false-positive violations."

This is then a problem in the maven-pmd-plugin - but I usually don't run pmd:check, I simply always execute the phase "verify" - which takes care of all this. Because I don't want to run pmd, but I just want to verify that my project passes the project's quality goals.

The missing classes/compilation phase might be handled smarter. There is similar case with the [typeresolution + aggregrate](https://maven.apache.org/plugins/maven-pmd-plugin/faq.html#typeresolution_aggregate) problem.

Report any issues regarding maven-pmd-plugin there: https://issues.apache.org/jira/projects/MPMD/summary

Btw. - I didn't find the auxclasspath in your [pmdxml.groovy](https://github.com/Fernal73/DSAlgos/blob/master/pmdxml.groovy).

@Fernal73  So, in summary, there are no false-positives then?

> I'm not sure CloneThrowsCloneNotSupportedException is useful, as calling super.clone() already forces you to declare the 
> exception or handle it... why does it exist?

Don't know - that's a very old rule. Object.clone() throws this checked Exception, but you don't need to declare that when you override the clone method. You might need to catch it though when calling super.clone(). Maybe a candidate for retirement?

**Comment by Fernal73** [comment]
> 
> > I'm not sure CloneThrowsCloneNotSupportedException is useful, as calling super.clone() already forces you to declare the
> > exception or handle it... why does it exist?
> 
> Don't know - that's a very old rule. Object.clone() throws this checked Exception, but you don't need to declare that when you override the clone method. You might need to catch it though when calling super.clone(). Maybe a candidate for retirement?

I think I set up my pom to run PMD in the verify phase after doing a lot of digging for auxclasspath. But it's been quite a while and it's difficult to recall what was done months or years ago without checking back with the configuration.

I can't answer for the justification of the existing PMD rules. My point is that the CloneNotSupportedException has to be thrown or caught. The Java compiler supports both. So should the PMD rules.

Personally, given that clone is problematic, Copyconstrutors are recommended instead.
 I simply wonder if there's any performance penalty involved.

**Comment by Fernal73** [comment]
Wrt to the [linked warning](https://maven.apache.org/plugins/maven-pmd-plugin/faq.html#typeresolution_aggregate):
'What does the warning "The project xyz does not seem to be compiled. PMD results might be inaccurate." mean?'

I don't think the Maven plugin checks if source files have been modified since the last build. That will have to be a suggestion for improvement to the Maven plugin team.

**Comment by Fernal73** [comment]
Wrt to the rules written in  Xpath , these can be converted to java rules. But this is probably already on your plate for PMD 7.0.0.

**Comment by Fernal73** [comment]
I think I'd suggested earlier that PMD have some way of identifying which rules need type resolution and which don't. If that can't be provided in the documentation, it can be specified in the violation message with some sort of marker or additional field. 
Sometimes, I think I would like to run PMD prior to compilation (like Checkstyle but after formatting) but only for the rules that don't require type resolution and the ones that do after compilation. 
Also, if there's more byte code analysis happening with PMD 7.0.0, how close is PMD getting to Spotbugs which analyses only the class files?

**Comment by Fernal73** [comment]
> Don't know - that's a very old rule. Object.clone() throws this checked Exception, but you don't need to declare that when you override the clone method. You might need to catch it though when calling super.clone(). Maybe a candidate for retirement?

CloneNotSupportedException is not a runtime exception. So it has to be caught or thrown. 
IMO, unless Oracle retires the clone method, there's no scope for retiring this rule. You may have a different understanding.

**Comment by Fernal73** [comment]
> Btw. - I didn't find the auxclasspath in your [pmdxml.groovy](https://github.com/Fernal73/DSAlgos/blob/master/pmdxml.groovy).
> 
The auxclasspath is not needed since I'm not executing this on java source but only XML files. What I am more concerned with is running the profiles checkstyleg and pmdxml together which causes in a clash in the ANTLR versions used by the latest versions of Checkstyle and PMD. 

Message from Checkstyle

```text
Backed up suppressions-xpath.xml
ANTLR Tool version 4.9.1 used for code generation does not match the current runtime version 4.7.2ANTLR Runtime version 4.9.1 used for parser compilation does not match the current runtime version 4.7.2ANTLR Tool version 4.9.1 used for code generation does not match the current runtime version 4.7.2ANTLR Runtime version 4.9.1 used for parser compilation does not match the current runtime version 4.7.2
```

**Comment by Fernal73** [comment]
> 
> The missing classes/compilation phase might be handled smarter. There is similar case with the [typeresolution + aggregrate](https://maven.apache.org/plugins/maven-pmd-plugin/faq.html#typeresolution_aggregate) problem.
> 
> Report any issues regarding maven-pmd-plugin there: https://issues.apache.org/jira/projects/MPMD/summary
> 

@adangel What do you think should go in a Jira report? How can the plugin be enhanced?

**Comment by oowekyala** [comment]
This discussion is all over the place

I think @Fernal73 you make some suggestions that should be directed to the maven plugin. Please open tickets there and move the relevant parts of the discussion

> @adangel What do you think should go in a Jira report? How can the plugin be enhanced?

You'd like to change the warning messages, right? Also you write this:

> I don't think the Maven plugin checks if source files have been modified since the last build. That will have to be a suggestion for improvement to the Maven plugin team.

---

> Also, if there's more byte code analysis happening with PMD 7.0.0, how close is PMD getting to Spotbugs which analyses only the class files?

https://github.com/pmd/pmd/discussions/new

--- 

> CloneNotSupportedException is not a runtime exception. So it has to be caught or thrown.
IMO, unless Oracle retires the clone method, there's no scope for retiring this rule. You may have a different understanding.

> Yes, the rule can be retired. It's not very nice to have to wrap a call to an object's clone method with a try-catch block especially when CloneNotSupportedException is thrown only if the Cloneable interface is not implemented.

This exemplifies how this discussion is derailing.. Please focus. Retiring the rule could be described somewhere else.

---

I think it's safe to say the original ticket may be closed as there are no FPs here

**Comment by Fernal73** [comment]
> 
> I think @Fernal73 you make some suggestions that should be directed to the maven plugin. Please open tickets there and move the relevant parts of the discussion
> 
> > @adangel What do you think should go in a Jira report? How can the plugin be enhanced?
> 
> You'd like to change the warning messages, right? Also you write this:
> 

I can put those all in, no problem. What I'd like to know is if @adangel believes that those will be enough to make the report actionable i.e., will the issues be fixed? Previous history suggests otherwise.


> I think it's safe to say the original ticket may be closed as there are no FPs here.

I'd put CloneReturnTypeMustMatchClassName bug to incomplete requirements. It's hardly likely that I will change my above code sample to meet this rule.
